thrift.js

```diff
@@ -117,37 +118,112 @@ function Thrift(options) {
     this.allowIncludeAlias = options.allowIncludeAlias || false;
     this.allowOptionalArguments = options.allowOptionalArguments || false;
 
-    this.filename = options.entryPoint;
+    if (options.entryPoint) {
+        return this.load(options.entryPoint);
+    }
+}
+
+Thrift.prototype.models = 'module';
+
+Thrift.prototype.Message = Message;
+
+Thrift.prototype.load = function load(entryPoint) {
+    this.memo[entryPoint] = this;
+    this.filename = entryPoint;
     this.dirname = path.dirname(this.filename);
-    this.memo[this.filename] = this;
 
-    var ast = this.asts[options.entryPoint];
-    var source = this.idls[options.entryPoint];
-    if (!source && !ast) {
+    this.exception = null;
+
+    if (this.asyncFetch) {
+        var model = this;
+        return this._buildAst(this.filename, this.allowIncludeAlias)
+            .then(function () { model._compileAndLink(); return model; });
+    }
+
+    this._buildAst(this.filename, this.allowIncludeAlias);
+    this._compileAndLink();
+    return this;
+}
+
+Thrift.prototype._buildAst = function _buildAst(filename, allowIncludeAlias, alreadyBuilt) {
```

Comment:
`allowIncludeAlias` and `alreadyBuilt` should not need to be passed as arguments since they’re shared state of the transitive closure of `Thrift` objects.

Perhaps we should have a `this.parsed` map that we share among all `Thrift` objects?

Would that be redundant with presence in the `this.asts` map? It seems to me that being non-null there should be synonymous with having been parsed.

I do fully expect that we need `loaded` and `loading` maps. In that past, I’ve needed to distinguish the two to prevent duplicate effort.

Reply:
Thanks again for your feedback. In the process of preparing the required changes, I would like to clarify something. `allowIncludeAlias` and `alreadyBuilt` were passed as argument because the included sources were fetched before any other `Thrift` object creation. We could share the `alreadyBuilt` map, but the `allowIncludeAlias` argument has a different purpose here (it is set to always `true` from the root's children on).

Because parsing is done before compiling, we don't build any other `Thrift` model in the potentially asynchronous part of the code. This piece of code protects from fetching a source twice, assuming it's atomic :
```javascript
if (alreadyBuilt[filename]) {
  return ;
}
alreadyBuilt[filename] = true;
```

Is this safe enough according to you, or the possibility for the piece of code above to be executed twice with the same `filename` argument at the same time should be accounted for?

Reply:
Actually, is it even possible that the code above is executed twice at the same time? I don't know JavaScript very well (be it the spec or the implementations), but if I understand correctly, JavaScript engines won't run a "thread" until the one currently being executed is fully done executing. In our case, that means that the piece of code I shared is enough to guarantee that an async fetch could happen only once per `filename`, as there can't be a context switch between the `if` condition and `alreadyBuilt[filename] = true`.

Is that a correct interpretation? If not, could you expand a little on the `loaded` and `loading` maps you mentioned? I'm not familiar with these.